### PR TITLE
[flink] Introduce partition mark done with batch

### PR DESCRIPTION
**Pull request by wwj6591812**
### Purpose

Base on https://github.com/apache/paimon/pull/3317

This PR is intended to support a mechanism for marking a partition as finished and subsequently pulling up offline scheduling jobs with flink batch job.

### CODE REVIEW

File: docs/layouts/shortcodes/generated/flink_connector_configuration.html [hunk 1]
```diff
@@ -98,6 +98,12 @@
             <td>String</td>
             <td>Action to mark a partition done is to notify the downstream application that the partition has finished writing, the partition is ready to be read.<br />1. 'success-file': add '_success' file to directory.<br />2. 'done-partition': add 'xxx.done' partition to metastore.<br />Both can be configured at the same time: 'done-partition,success-file'.</td>
         </tr>
+        <tr>
+            <td><h5>partition.mark-done-with-batch</h5></td>
```
Comment (JingsongLi) [inline docs/layouts/shortcodes/generated/flink_connector_configuration.html:102]: `partition.end-input-to-done`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.flink.api.common.state.OperatorStateStore;
+
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.paimon.flink.FlinkConnectorOptions.PARTITION_IDLE_TIME_TO_DONE;
+import static org.apache.paimon.flink.FlinkConnectorOptions.PARTITION_TIME_INTERVAL;
+
+/** A Abstract class of PartitionMarkDoneTrigger. */
+public abstract class AbstractPartitionMarkDoneTrigger {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java:35]: just name `PartitionMarkDoneTrigger`.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/BatchPartitionMarkDoneTrigger.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package org.apache.paimon.flink.sink.partition;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/** Trigger to mark partitions done with batch job. */
+public class BatchPartitionMarkDoneTrigger extends AbstractPartitionMarkDoneTrigger {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/BatchPartitionMarkDoneTrigger.java:26]: implement an end-input?

Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/BatchPartitionMarkDoneTrigger.java:26]: EndInputPartitionMarkDoneTrigger

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        notifyPartition(partition, System.currentTimeMillis());
+    }
+
+    public void notifyPartition(String partition, long currentTimeMillis) {
+        if (!StringUtils.isNullOrWhitespaceOnly(partition)) {
+            this.pendingPartitions.put(partition, currentTimeMillis);
+        }
+    }
+
+    public List<String> donePartitions() {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java:53]: You can introduce a `List<String> donePartitions(boolean endInput)`.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java (hunk 2 above)
Comment (wwj6591812) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/BatchPartitionMarkDoneTrigger.java (hunk 3 above)
Comment (wwj6591812) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/BatchPartitionMarkDoneTrigger.java]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java (hunk 4 above)
Comment (wwj6591812) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java]: done

File: docs/layouts/shortcodes/generated/flink_connector_configuration.html (hunk 1 above)
Comment (wwj6591812) [reply docs/layouts/shortcodes/generated/flink_connector_configuration.html]: done

**Comment by JingsongLi** [comment]
We can unify batch and streaming to one trigger.

**Comment by wwj6591812** [comment]
yes, updated，pls cc, thx

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**